### ISSUE TITLE
Random Game Crash/Freeze in Ocarina of Time with Dynamic Recompiler

### ISSUE THREAD

**Issue by flagrama** [libretro/mupen64plus-libretro-nx#180]
This is just a proper GitHub issue for the random game crash/freeze that happens in Ocarina of Time on the dynamic_recompiler that was noticed in the OoT Randomizer. I hadn't created an issue prior to this, as there wasn't any evidence that this was strictly an emulator issue, it could have been something our code was doing that was interacting with OoT in an odd way.

However, the freeze seems to happen in vanilla 1.0 NTSC (CRCs: CD16C529 EC95702D): https://imgur.com/a/4gIhtsB 
![image](https://user-images.githubusercontent.com/634079/80550080-e34b2180-898c-11ea-87ce-57623a050b9c.png)

and in vanilla 1.2 NTSC (CRCs: 32120C23 3C7DBF5B): https://imgur.com/a/ZGnbQyG
![image](https://user-images.githubusercontent.com/634079/80550093-eba35c80-898c-11ea-987d-8b74026b9461.png)

The 1.0 version is virtually identical to the same freeze in the randomizer.

This issue does not appear to occur with the cached_interpreter.

**Comment by Pearate** [comment]
I had a crash in vanilla 1.0 NTSC (CRC32 Checksum: CD16C529)  with the cached_interpreter.
I changed to pure_interpreter and I had no crash so far (2 hour of play) 
I keep you informed.

**Comment by fig02** [comment]
I dont suppose that this information will help you guys figure things out on the emulator side of things, but I took these notes a little while ago on why this crash was happening on the game side: https://pastebin.com/Uf0XzP6U

Tldr some code that is expecting a pointer to a horse is running when it shouldn't be. A bit in Links state flags that denotes he is riding a horse is set when it shouldn't be, and the code dereferences a NULL pointer to the actor he is "riding".

**Comment by Pearate** [comment]
I just had a crash with the pure_interpreter too. So I can say that no interpreter works with OoT.
If that's a clue, all of the crashes ware in the adult world.
I think I will play this game with another emulator.
Thx

**Comment by m4xw** [comment]
> 
> 
> I just had a crash with the pure_interpreter too. So I can say that no interpreter works with OoT.
> If that's a clue, all of the crashes ware in the adult world.
> I think I will play this game with another emulator.
> Thx

Show your core options

**Comment by Pearate** [comment]
Here is my option file.
[Legend of Zelda, The - Ocarina of Time (USA).zip](https://github.com/libretro/mupen64plus-libretro-nx/files/4554804/Legend.of.Zelda.The.-.Ocarina.of.Time.USA.zip)
Any suggestion is welcome, I find it a little labyrinthine.

**Comment by m4xw** [comment]
To rule out parallel rsp (tho unlikely), I would like to ask you to give it a try with cached interpreter + HLE, since HLE is my primary testing case.
Also another very unlikely case could be "EnableNativeResTexrects".
At least you are the first person to report issues with interpreters.
What are your system specs?
Rest of your config looks fine on first glance.

**Comment by m4xw** [comment]
Also wouldnt be a first if LLE would be "too accurate", heh.

**Comment by Pearate** [comment]
> What are your system specs?

Windows 10 
i7 - 9700K @ CPU 3.60 GHz  
RAM 16  Go
NVIDIA GeForce RTX 2070

I try tomorrow with :
"cached interpreter" + "HLE"
"EnableNativeResTexrects" -> off

What do you mean by 'if LLE would be "too accurate"' ?

**Comment by Testsr** [comment]
If LLE did not emulate a bug that the rom hack required.

**Comment by flagrama** [comment]
This isn't happening in (just) a ROM hack. This issue was created due to it happening in a retail release.

**Comment by Pearate** [comment]
Hi,
I tried to play in the temple of water with this option:
"cached interpreter" + "HLE"
"EnableNativeResTexrects" -> off

But I stopped because I had constant sound glitch + slow down.
Same with pure_interpreter...

Then I played with:
"Dynamic_recompiler" + "HLE"
"EnableNativeResTexrects" -> off

I had a freeze again the doppelganger boss fight.
Here is the debug: 
![freeze](https://user-images.githubusercontent.com/35386725/81509194-387e2000-9309-11ea-8621-be100842c6c7.jpg)

I join a save stat too  in case.
[Legend of Zelda, The - Ocarina of Time (USA).zip](https://github.com/libretro/mupen64plus-libretro-nx/files/4606665/Legend.of.Zelda.The.-.Ocarina.of.Time.USA.zip)

Something intrigues me, not everyone is facing bugs with this game apparently.
Why ?  What are there settings ?

**Comment by m4xw** [comment]
Someone give this a shot: [mupen64plus_next_libretro.zip](https://github.com/libretro/mupen64plus-libretro-nx/files/4609187/mupen64plus_next_libretro.zip)

**Comment by m4xw** [comment]
In case you get "failed to open libretro core": https://git.m4xw.net/libretro/cores/mupen64plus-libretro-nx/-/jobs/7728/artifacts/browse

**Comment by Pearate** [comment]
> Someone give this a shot: [mupen64plus_next_libretro.zip](https://github.com/libretro/mupen64plus-libretro-nx/files/4609187/mupen64plus_next_libretro.zip)

Nothing has changed with this new core. What was expected ? 

I did several tests, only "cached interpreter" + "HLE" seems to work without freezing or random crash. But in the same way, this parameters causes a sound problem + a slowdown.
What can I do to fix the sound and the slowdown ?

**Comment by m4xw** [comment]
For now added the good ol' skip tlb option in ba73792f43ccc20a40b5317e77a15195daa69128 on the rebase branch.
Gonna leave the issue open tho, we are still on it but it ends up a giant drag to debug.
Fixed a bunch of issues but none of it was the cause :/
Fwiw skip tlb will be useful when GdbStub comes in too

**Comment by jacksonff** [comment]
I've been getting the same random freezing issue in Master Quest. But found this video https://www.youtube.com/watch?v=ojVXc3IiPxo (min 15:15), where he says to enable "SRAM autosave interval" from RetroArch settings in order to avoid issues. This is the default option so my troubles weren't caused by that. You should try a ROM hack called "Petrie's Challenge", where same kind of freezes happen more often.

It may be some kind of placebo effect, but I wasn't getting any freezes if I kept creating save states or saving manually. So I disabled the option and it hasn't froze anymore (by now). I'm using latest RetroArch (1.9.0), mupen64plus-next v2.1 (ea1c677), dynarec, HLE and GlideN64. Hoping this comment is not totally useless.

**Comment by m4xw** [comment]
This sounds unlikely, but if it pokes it at a bad time... hmmm...

**Comment by Bowi-QC** [comment]
Is this the issue I had too?

**Comment by Bowi-QC** [comment]
> For now added the good ol' skip tlb option in [ba73792](https://github.com/libretro/mupen64plus-libretro-nx/commit/ba73792f43ccc20a40b5317e77a15195daa69128) on the rebase branch.
> Gonna leave the issue open tho, we are still on it but it ends up a giant drag to debug.
> Fixed a bunch of issues but none of it was the cause :/
> Fwiw skip tlb will be useful when GdbStub comes in too

Which mode should I select for OoT?

**Comment by m4xw** [comment]
"Skip if not using"

**Comment by Bowi-QC** [comment]
Uh.. Autechre made me doubt about something...

Does this bug depend on the RDP and RSP plugins one uses or does it happen with any plugin as long as Dynarec is chosen?

(I was using ParaLLEl RSP and RDP each time it happened.)

**Comment by m4xw** [comment]
its a mupen64plus-core issue.

**Comment by flagrama** [comment]
Should this issue be duplicated there and link it to this one?

**Comment by m4xw** [comment]
If People want to test: https://git.libretro.com/libretro/mupen64plus-libretro-nx/-/jobs/247457/artifacts/browse

**Comment by m4xw** [comment]
Success so far 👀

**Comment by fpi1337** [comment]
> Success so far 👀

It's been a while but I just came along to say that this still is an issue. Experienced some freezes the last days using the newest core I built from source.

(Den Spaß hatten wir ja mal vor einiger Zeit bei Majora's Mask, aber scheinbar macht der Dynarec immer noch Probleme :P)

**Comment by m4xw** [comment]
Its not on develop

**Comment by m4xw** [comment]
@fpi1337 current wip is here https://git.libretro.com/libretro/mupen64plus-libretro-nx/-/commits/shenanigans
Gillou made a bunch of changes, including fixing it for other archs than amd64, hence the delay

**Comment by fpi1337** [comment]
You mean the one from April? If that's the case, I'll try it out and build it myself.



Offtopic und auf Deutsch, da leichter zu veranschaulichen: Ich hatte im Übrigen auch neulich wieder auf meinem Pi 4 einen Freeze bei Majora's Mask. Gleiches Szenario wie damals und vermutlich auch wieder Dynarec. Habe die Ignore TLB-Option allerdings nicht aktiviert gehabt. 

Ferner habe ich mit dem Core enorme Audio-Stutterings (unter anderem auf meinem alten Sony Vaio-Laptop, der beispielsweise mit dem ParaLLEL-Core (ohne Angrylion versteht sich) OoT ohne Probleme emuliert). Linux wohlgemerkt. Ich habe bereits nach diversen Option zur Audioanpassung gesucht, allerdings scheint Dein Core keinerlei Optionen zur Anpassung (bspws. Audiosynch/stretching) zu bieten oder übersehe ich da etwas?

**Comment by flagrama** [comment]
The 10 year old bug is with the new dynarec right? Or did it affect old dynarec as well, or not even a dynarec issue at root? I know old dynarec builds never seem to run into the issue, and Next is pretty much the only public build that uses the new dynarec.

**Comment by m4xw** [comment]
> 
> 
> The 10 year old bug is with the new dynarec right? Or did it affect old dynarec as well, or not even a dynarec issue at root? I know old dynarec builds never seem to run into the issue, and Next is pretty much the only public build that uses the new dynarec.

Its new_dynarec only

**Comment by fpi1337** [comment]
> Das war damals ein interrupt issue

Nun, ich hatte auf dem Pi 4 vor 'ner Woche bei Majoras Mask wieder einen Freeze (könnte aber daran liegen, dass ich die ROM nach der Installation direkt auf die SD-Karte gepackt hatte. Ocarina of Time statete unter diesen Umständen nicht mal. Allerdings ließ sich das durch Kopieren der ROMS auf ein externes Speichermedium beheben.) - kurz nachdem ich die Bomber gefangen hatte. Ist sowas in der Art wieder aufgetreten bzw. bekannt? Ich hatte ja damals Deine gefixten Cores ein Stück weit mit ausprobiert und dort Rückmeldung gegeben, dass mir nichts mehr in der Art untergekommen ist.

> Alles retroarch einstellungen

Muss mich dahingehend korrigieren. Habe mir nochmal für den Standalone Mupen64Plus-Emulator GlideN64 kompiliert und habe da einen ähnlichen Effekt. Es liegt also eher daran, dass mein Laptop zu alt ist und mit GlideN64 nicht klarkommt (vermutlich unterstützt die alte Intel HD-Grafikkarte noch kein Open GL 3.3). Auf meinem Hauptrechner funktioniert alles ohne Audio-Probleme.

> Der Dynarec bug ist über 10 jahre alt!

Dann sollten wir ihm wohl einen Strauß Blumen zum Jubiläum schicken. ;)

Eine kleine Frage noch am Rande, dann ist auch Schluss mit Deutsch:

Was hat es mit den TLB-Exception-Option genau auf sich? Verhindert die Option den Ocarina of Time-Freeze unter Dynarec? Und welche Einstellung ist dort die richtige: "Always Ignore TLB Exceptions" oder "Ignore TLB Exceptions if not using TLB"?

**Comment by m4xw** [comment]
BEGONE
